### PR TITLE
Add indexer to fakeCilent when bumping k8s.io deps to v0.26.1 

### PR DESCRIPTION
**Pull request by kerthcet**
#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/kueue/pull/586

#### Special notes for your reviewer:

The test error was introduced in https://github.com/kubernetes-sigs/controller-runtime/pull/2025, for fakeClient will check the indexes but we didn't register any in unit tests.

But we can register these indexes when building the client.

### CODE REVIEW

**Review by tenzen-y — COMMENTED** [review]
/lgtm

**Comment by alculquicondor** [comment]
can you bump all the k8s libraries together?

**Comment by kerthcet** [comment]
Failed test: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/588/pull-kueue-test-unit-main/1628060447326343168

**Comment by kerthcet** [comment]
The test error was introduced in https://github.com/kubernetes-sigs/controller-runtime/pull/2025, for fakeClient will check the indexes but we didn't register any in unit tests.

**Comment by kerthcet** [comment]
But we can register these indexes when building the client.

File: pkg/util/testing/client.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha2"
+	"sigs.k8s.io/kueue/pkg/constants"
+	utilindexer "sigs.k8s.io/kueue/pkg/util/indexer"
+)
+
+func NewFakeClient(objs ...client.Object) client.Client {
```
Comment (kerthcet) [inline pkg/util/testing/client.go:30]: Main functions located at this file.

File: pkg/util/indexer/indexer.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package indexer
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha2"
+)
+
+var (
+	IndexQueueClusterQueue = func(obj client.Object) []string {
```
Comment (kerthcet) [inline pkg/util/indexer/indexer.go:26]: These index functions both used in package `cache` and `queue`, so moved to `util/indexer`.

**Comment by kerthcet** [comment]
Yes, the test error fixed.

File: pkg/util/indexer/indexer.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (tenzen-y) [inline pkg/util/indexer/indexer.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: pkg/constants/constants.go [hunk 4]
```diff
@@ -45,3 +47,9 @@ const (
 	// marked as default.
 	DefaultPriority = 0
 )
+
+const (
+	MatchingFieldWorkloadQueueKey        = "spec.queueName"
+	MatchingFieldQueueClusterQueueKey    = "spec.clusterQueue"
+	MatchingFieldWorkloadClusterQueueKey = "spec.admission.clusterQueue"
```
Comment (alculquicondor) [inline pkg/constants/constants.go:54]:
```suggestion
	WorkloadQueueKey        = "spec.queueName"
    WorkloadClusterQueueKey = "spec.admission.clusterQueue"
	QueueClusterQueueKey    = "spec.clusterQueue"
```

File: pkg/util/testing/client.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	scheme := runtime.NewScheme()
```
Comment (alculquicondor) [inline pkg/util/testing/client.go:31]: use `MustGetScheme`

File: pkg/util/indexer/indexer.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/util/indexer/indexer.go:26]: Maybe we should just have a single `SetupIndexes` function for the kueue APIs.

File: pkg/util/testing/client.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha2"
+	"sigs.k8s.io/kueue/pkg/constants"
+	utilindexer "sigs.k8s.io/kueue/pkg/util/indexer"
```
Comment (alculquicondor) [inline pkg/util/testing/client.go:27]:
```suggestion
	"sigs.k8s.io/kueue/pkg/util/indexer"
```

File: pkg/constants/constants.go (hunk 4 above)
Comment (alculquicondor) [reply pkg/constants/constants.go:54]: Also, just put then in the `indexer` package.

File: pkg/util/testing/client.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+		panic(err)
+	}
+
+	return fake.NewClientBuilder().WithScheme(scheme).
+		WithIndex(&kueue.LocalQueue{}, constants.MatchingFieldQueueClusterQueueKey, utilindexer.IndexQueueClusterQueue).
+		WithIndex(&kueue.Workload{}, constants.MatchingFieldWorkloadQueueKey, utilindexer.IndexWorkloadQueue).
+		WithIndex(&kueue.Workload{}, constants.MatchingFieldWorkloadClusterQueueKey, utilindexer.IndexWorkloadClusterQueue).
+		WithObjects(objs...).
+		Build()
+}
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/util/testing/client.go:45]: No, we should add scheme first.

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/util/testing/client.go:45]: I think we can make processes to add schemes common. WDYT?

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/util/testing/client.go:45]: Like the following:

```go
func NewFakeClient(objs ...client.Object) client.Client {
        return NewClientBuilder().WithObjects(objs...).Build()
}

func NewClientBuilder() *fake.ClientBuilder {
	scheme := runtime.NewScheme()
	if err := corev1.AddToScheme(scheme); err != nil {
		panic(err)
	}
	if err := kueue.AddToScheme(scheme); err != nil {
		panic(err)
	}

	return fake.NewClientBuilder().WithScheme(scheme).
		WithIndex(&kueue.LocalQueue{}, constants.MatchingFieldQueueClusterQueueKey, utilindexer.IndexQueueClusterQueue).
		WithIndex(&kueue.Workload{}, constants.MatchingFieldWorkloadQueueKey, utilindexer.IndexWorkloadQueue).
		WithIndex(&kueue.Workload{}, constants.MatchingFieldWorkloadClusterQueueKey, utilindexer.IndexWorkloadClusterQueue)
}
```

File: pkg/util/testing/client.go (hunk 5 above)
Comment (kerthcet) [reply pkg/util/testing/client.go:31]: I think we can remove MustGetScheme now.

File: pkg/util/testing/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/util/testing/client.go:45]: That's right

File: pkg/util/indexer/indexer.go (hunk 2 above)
Comment (kerthcet) [reply pkg/util/indexer/indexer.go:32]: Do you mean have only one `setupIndexes()` in main.go? Not separated into different packages like `queue`, `cache`, `job`.

**Comment by kerthcet** [comment]
All addressed except https://github.com/kubernetes-sigs/kueue/pull/588#discussion_r1114570049.

File: pkg/util/indexer/indexer.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/util/indexer/indexer.go:32]: probably not in main.go, because the tests need it. But yes, a single place.

**Review by tenzen-y — COMMENTED** [review]
Thanks!
LGTM from my side.

File: pkg/util/indexer/indexer.go (hunk 2 above)
Comment (kerthcet) [reply pkg/util/indexer/indexer.go:32]: Let me make this cleanup tomorrow. We can leave this a follow up if that's ok with you.

File: pkg/util/indexer/indexer.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+var (
+	IndexQueueClusterQueue = func(obj client.Object) []string {
+		q, ok := obj.(*kueue.LocalQueue)
+		if !ok {
+			return nil
+		}
+		return []string{string(q.Spec.ClusterQueue)}
+	}
+
+	IndexWorkloadQueue = func(obj client.Object) []string {
```
Comment (alculquicondor) [inline pkg/util/indexer/indexer.go:40]:
```suggestion
	func IndexWorkloadQueue(obj client.Object) []string {
```

why overcomplicate things :)

File: pkg/util/indexer/indexer.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/util/indexer/indexer.go:32]: sg

**Review by alculquicondor — COMMENTED** [review]
/approve

File: pkg/util/indexer/indexer.go (hunk 8 above)
Comment (alculquicondor) [reply pkg/util/indexer/indexer.go:40]: The follow up PR will just move the entire SetupIndexes functions here.

Comment (tenzen-y) [reply pkg/util/indexer/indexer.go:40]: sgtm.

**Comment by alculquicondor** [comment]
@kerthcet we can also get rid of checks like this:
https://github.com/kubernetes-sigs/kueue/blob/c45d3dd98e3cae593e844092e2f334879a40ce7c/pkg/queue/manager.go#L172

**Pull request merged**